File: ts/src/env.ts
/**
 * Loads `.env` from the working directory (or DISPATCHER_ENV_FILE) into process.env
 * without overriding variables that are already set. Values are never logged.
 */
import { existsSync } from "node:fs";
import { dirname, resolve } from "node:path";

/** Variables that loadEnv() itself added to process.env, so wrappers can keep them from child CLIs. */
export const loadedFromEnvFile: string[] = [];

/** Looks in the working directory, then up to three parents (the repo root when run from ts/). */
export function loadEnv(file = process.env.DISPATCHER_ENV_FILE ?? ".env"): boolean {
  let dir = process.cwd();
  for (let i = 0; i < 4; i++) {
    const path = resolve(dir, file);
    if (existsSync(path)) {
      try {
        const before = new Set(Object.keys(process.env));
        // Node 20.12+ / 22: parses the file and only sets variables not already defined.
        process.loadEnvFile(path);
        for (const k of Object.keys(process.env)) if (!before.has(k)) loadedFromEnvFile.push(k);
        return true;
      } catch {
        return false;
      }
    }
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return false;
}

/** Env names dispatcher reads. Only presence is ever reported, never values. */
export const ENV = {
  typesafe: "TYPESAFE_API_KEY",
  anthropic: "ANTHROPIC_API_KEY",
  openai: "OPENAI_API_KEY",
  openaiBaseURL: "OPENAI_BASE_URL",
  openrouter: "OPENROUTER_API_KEY",
  port: "DISPATCHER_PORT",
  apiKey: "DISPATCHER_API_KEY",
  baseline: "DISPATCHER_BASELINE",
  fallback: "DISPATCHER_FALLBACK",
  routeAll: "DISPATCHER_ROUTE_ALL",
  ledger: "DISPATCHER_LEDGER_PATH",
  catalog: "DISPATCHER_CATALOG",
  jevTimeout: "DISPATCHER_JEV_TIMEOUT_MS",
} as const;

export function has(name: string): boolean {
  const v = process.env[name];
  return typeof v === "string" && v.trim().length > 0;
}

File: csharp/Dispatcher/Env.cs
namespace LlmDispatcher;

/// <summary>
/// Loads a <c>.env</c> file into the process environment without overriding variables that are
/// already set. Looks in the working directory, then up to three parents. Values are never logged.
/// </summary>
public static class Env
{
    public const string TypeSafe = "TYPESAFE_API_KEY";
    public const string Anthropic = "ANTHROPIC_API_KEY";
    public const string OpenAI = "OPENAI_API_KEY";
    public const string OpenAIBaseUrl = "OPENAI_BASE_URL";
    public const string OpenRouter = "OPENROUTER_API_KEY";
    public const string Port = "DISPATCHER_PORT";
    public const string ApiKey = "DISPATCHER_API_KEY";
    public const string Baseline = "DISPATCHER_BASELINE";
    public const string Fallback = "DISPATCHER_FALLBACK";
    public const string RouteAll = "DISPATCHER_ROUTE_ALL";
    public const string LedgerPath = "DISPATCHER_LEDGER_PATH";
    public const string Catalog = "DISPATCHER_CATALOG";
    public const string JevTimeout = "DISPATCHER_JEV_TIMEOUT_MS";

    public static bool Load(string? file = null)
    {
        file ??= Environment.GetEnvironmentVariable("DISPATCHER_ENV_FILE") ?? ".env";
        var dir = Directory.GetCurrentDirectory();
        for (var i = 0; i < 4 && dir is not null; i++)
        {
            var path = Path.Combine(dir, file);
            if (File.Exists(path))
            {
                Apply(File.ReadAllLines(path));
                return true;
            }
            dir = Path.GetDirectoryName(dir);
        }
        return false;
    }

    internal static void Apply(IEnumerable<string> lines)
    {
        foreach (var raw in lines)
        {
            var line = raw.Trim();
            if (line.Length == 0 || line.StartsWith('#')) continue;
            if (line.StartsWith("export ", StringComparison.Ordinal)) line = line[7..].TrimStart();
            var eq = line.IndexOf('=');
            if (eq <= 0) continue;
            var key = line[..eq].Trim();
            var value = line[(eq + 1)..].Trim();
            if (value.Length >= 2 && (value[0] == '"' && value[^1] == '"' || value[0] == '\'' && value[^1] == '\''))
                value = value[1..^1];
            else
            {
                var hash = value.IndexOf(" #", StringComparison.Ordinal);
                if (hash >= 0) value = value[..hash].TrimEnd();
            }
            if (string.IsNullOrEmpty(Environment.GetEnvironmentVariable(key)))
                Environment.SetEnvironmentVariable(key, value);
        }
    }

    /// <summary>Presence only. Never returns or prints the value.</summary>
    public static bool Has(string name) => !string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable(name));

    public static string? Get(string name) => Environment.GetEnvironmentVariable(name);
}
